lib/mayaUsd/resources/scripts/mayaUsdAddMayaReference.mel

```diff
@@ -629,12 +993,159 @@ global proc string addMayaReferenceToUsd(string $ufePath)
+}
+
+global proc usdMayaRef_changeVariantSetName(string $value)
+{
+    global string $gUsdMayaReferenceUfePath;
+
+    string $createNew = getMayaUsdLibString("kMayaRefCreateNew");
+
+    // If the "Create New" item is selected, we show the text input
+    // to the right of the optionMenu. For existing variant set names
+    // the field is hidden.
+    int $showField = ($value == $createNew);
+    textField -e -visible $showField usdMayaReferenceVariantSetNameTextField;
+
+    // When the variant set name is showing "Create New", the only option
+    // in the variant name menu is also "Create New".
+    optionMenu -edit -visible true -deleteAllItems usdMayaReferenceVariantNameOptionMenu;
+    textField -e -visible true usdMayaReferenceVariantNameTextField;
+    menuItem -label $createNew -parent usdMayaReferenceVariantNameOptionMenu;
+
+    // If we picked a new variant set name, populate the variant name optionMenu
+    // with all the variant names (after the "Create New" menu item).
+    if (! $showField)
+    {
+        string $savedVarName = `optionVar -q usdMayaReferenceVariantNameOption`;
+        string $restoreVarName = "";
+
+        string $variantNames[] = `python("import mayaUsdAddMayaReference; mayaUsdAddMayaReference.getVariantNames('" + $gUsdMayaReferenceUfePath + "', '" + $value + "')")`;
+        for ($varName in $variantNames)
+        {
+            menuItem -label $varName -parent usdMayaReferenceVariantNameOptionMenu;
+            if ($varName == $savedVarName) {
+                $restoreVarName = $varName;
+            }
+        }
+
+        if ($restoreVarName != "") {
+            optionMenu -edit -value $restoreVarName usdMayaReferenceVariantNameOptionMenu;
+            // Since we restored a value in the option menu we need to hide the field.
+            textField -e -visible false usdMayaReferenceVariantNameTextField;
+        }
+    }
+}
+
+global proc usdMayaRef_changeVariantName(string $value)
+{
+    // If the "Create New" item is selected, we show the text input
+    // to the right of the optionMenu. For existing variant names
+    // the field is hidden.
+    string $createNew = getMayaUsdLibString("kMayaRefCreateNew");
+    int $vis = ($value == $createNew);
+    textField -e -visible $vis usdMayaReferenceVariantNameTextField;
+}
+
+global proc usdMayaRef_changeVariantSetNameText(string $value)
+{
+    // The text field cannot be empty. Reset to default value if it is.
+    if ($value == "") {
+        string $defValue = getMayaUsdLibString("kMayaRefDefaultVariantSetName");
+        textField -edit -text $defValue usdMayaReferenceVariantSetNameTextField;
+    } else {
+        // Make sure the name user entered doesn't contain any invalid characters.
+        string $validatedSetName = sanitizeName($value);
+        if ($value != $validatedSetName) {
+            textField -edit -text $validatedSetName usdMayaReferenceVariantSetNameTextField;
+        }
+    }
```

Comment:
And here - sanitize the variant set and variant names after user enters them, so the optionVar will have the correct value as well.